### PR TITLE
Cache mitigation

### PR DESCRIPTION
**Pull request by bigkevmcd**
## Problem

Too many secrets are created in parallel across nodes.
 
## Solution

This uses optimistic locking to prevent the secrets from persisting.

When a secret is created for an SA, the SA is annotated with the Secret reference, if this results in a conflict, the Secret is deleted.

This allows for the first process to annotate the resource to ensure that the secret exists.
 
### Automated Testing

* Test types added/modified:
    * Unit
    * Integration (Go Framework)
    * Integration (v2prov Framework)
    * Validation (Go Framework)
    * Other - Explain: _EXPLAIN_
    * None
    * _REMOVE NOT APPLICABLE BULLET POINTS ABOVE_
* If "None" - Reason: _EXPLAIN THE REASON_
  
* If "None" - GH Issue/PR: _LINK TO GH ISSUE/PR TO ADD TESTS_

Summary: _TODO_

### Regressions Considerations

_TODO_

Existing / newly added automated tests that provide evidence there are no regressions:
* _TODO_

### CODE REVIEW

File: tests/pkg/serviceaccounttoken/mitigation_test.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	typedcorev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/envtest"
+)
+
+func TestOptimisticLocking(t *testing.T) {
```
Comment (aruiz14) [inline tests/pkg/serviceaccounttoken/mitigation_test.go:20]: We could be nicer to our future selves who may forget why we did this and add some comments here.
In general, I like that for any test for a specific case, since they are not normally tested after the struct/method, and unlike table-driven tests, don't use a explicit name for each case.

File: tests/pkg/serviceaccounttoken/mitigation_test.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+	testEnv := &envtest.Environment{}
+	restCfg, err := testEnv.Start()
+	assert.NoError(t, err)
+	defer testEnv.Stop()
+
+	clientSet, err := kubernetes.NewForConfig(restCfg)
+	assert.NoError(t, err)
+	k8sClient := clientSet.CoreV1()
+
+	t.Run("creating a secret", func(t *testing.T) {
```
Comment (aruiz14) [inline tests/pkg/serviceaccounttoken/mitigation_test.go:30]: nit-pick: this does not provide much value?

File: pkg/serviceaccounttoken/secret.go [hunk 3]
```diff
@@ -204,29 +205,84 @@ func isSecretForServiceAccount(secret *corev1.Secret, sa *corev1.ServiceAccount)
...
-		return nil, fmt.Errorf("error ensuring secret for service account [%s:%s]: %w", sa.Namespace, sa.Name, err)
+		return nil, false, fmt.Errorf("getting updated service account %s: %w", keyFromObject(sa), err)
 	}
 
-	return secret, nil
+	return updated, true, nil
+
+}
+
+func keyFromObject(obj namespacedObject) namespacedName {
```
Comment (aruiz14) [inline pkg/serviceaccounttoken/secret.go:254]: nitpick: I don't see other usage for `keyFromObject` other than logging and error messages that would justify using a struct. In fact, the implementation of the `String()` method is primarily focused on that use case (due to wrapping around square brackets).
I see other places where we could use a similar approach (e.g. [line 227](https://github.com/rancher/rancher/pull/47273/files#diff-8a853a257f4c9e8e5812b706a4f6a56eaa32ff67e57f9ac5ac4f78f1e1659f04R227)) directly use string concatenation:
```
sa.Annotations[ServiceAccountSecretRefAnnotation] = secret.Namespace + "/" + secret.Name
```
What about renaming this to explicitly mention logging and make it directly return a string? (beware also that `GetNamespace()` could return an empty string for cluster level objects, while still matching the interface)

File: pkg/serviceaccounttoken/secret.go [hunk 4]
```diff
@@ -204,29 +205,84 @@ func isSecretForServiceAccount(secret *corev1.Secret, sa *corev1.ServiceAccount)
...
-	if secret != nil {
-		return secret, nil
+
+	return secret, nil
+}
+
+// returns true if the secret has changed.
+func annotateSAWithSecret(ctx context.Context, sa *corev1.ServiceAccount, secret *corev1.Secret, saClient clientv1.ServiceAccountInterface, secretClient clientv1.SecretInterface) (*corev1.ServiceAccount, bool, error) {
+	if sa.Annotations[ServiceAccountSecretRefAnnotation] != "" {
+		// TODO: I guess we should check to see if it's the existing SA?
```
Comment (aruiz14) [inline pkg/serviceaccounttoken/secret.go:221]: Unresolved TODO?

**Review by aruiz14 — DISMISSED** [review]
LGTM!

File: pkg/serviceaccounttoken/secret.go [hunk 5]
```diff
@@ -79,21 +63,38 @@ func EnsureSecretForServiceAccount(ctx context.Context, secretsCache corecontrol
...
 	}
 
 	if secret == nil {
-		secret, err = createServiceAccountSecret(ctx, sa, secretLister, secretClient, lockKey)
+		secret, err = createServiceAccountSecret(ctx, sa, secretLister, secretClient)
+		if err != nil {
+			return nil, err
+		}
+
+		sa, updated, err := annotateSAWithSecret(ctx, sa, secret, clientSet.CoreV1().ServiceAccounts(sa.Namespace), secretClient)
```
Comment (MbolotSuse) [inline pkg/serviceaccounttoken/secret.go:75]: I think there's still room for duplicates in this part of the code. The situation I'm thinking about is the following:
- Two Rancher replicas (or threads)
- Replica 1 receives a request, and updates the service account
- Replica 2 receives a request. It gets the new updated service account with the annotation.
- Since Replica 2 is using a cache, when it lists secrets through "ServiceAccountSecret", it doesn't see the new secret yet.
- Replica 2 creates a new secret (so there's now 2 secrets)
- `annotateSAWithSecret` sees that the annotation is on the service account, so it doesn't update the annotation, or delete the current secret.
- We use the new secret that we just made rather than the original one

Now future requests will eventually clean this up since ServiceAccountSecret will try to delete "duplicate" secrets. However, I think this could happen in ways that we don't prefer (such as by deleting a token before a thread can use it to make a call, returning an error to the user, or by deleting the secret listed on the service account, leaving the actual token that we are using different than what is listed on the secret).

High level, I think that this can be fixed by adjusting "ServiceAccountSecret" to take this new approach into account. More specifically, maybe we can have this read the annotation and extract the specific secret rather than doing a list and taking the first secret that belongs to this ServiceAccount. We'd still need to account for cleanup here of the existing duplicates, but maybe that's better done in a standalone fashion rather than evaluated on every request (or only evaluated when we add the annotation for the first time).

**Review by MbolotSuse — DISMISSED** [review]
Could be wrong, but I think that this leaves open at least one edge case where duplicate secrets could still be created. That being said, I think the overall theory should work if we can get this edge case and the cleanup figured out.

File: pkg/serviceaccounttoken/secret.go [hunk 6]
```diff
@@ -204,29 +233,89 @@ func isSecretForServiceAccount(secret *corev1.Secret, sa *corev1.ServiceAccount)
...
 	}
 
-	return secret, nil
+	return updated, true, nil
+
+}
+
+// This should be replaced with a types.NamespacedName which stringifies to
+// <namespace>/<name> but that would change the log output format.
+func logKeyFromObject(obj namespacedObject) namespacedName {
```
Comment (MbolotSuse) [inline pkg/serviceaccounttoken/secret.go:287]: Nit: I'm not sure that we need the extra overhead of this type for a few reasons:
- If we really want to have a separate function for the key due to re-use, then we should just return the output of the string function, rather than include the overhead of the string representation.
- I don't think that the exact interface of the key in the log statement is an API-level construct that we need backward compat for.

File: pkg/serviceaccounttoken/secret.go [hunk 7]
```diff
@@ -204,29 +233,89 @@ func isSecretForServiceAccount(secret *corev1.Secret, sa *corev1.ServiceAccount)
...
+		return updated, false, nil
+	}
+	if !apierrors.IsConflict(err) {
+		logrus.Debugf("Failed to update ServiceAccount for %s: %s", logKeyFromObject(sa), err)
+		return nil, false, err
+	}
+
+	// Rollback the optimistically created secret
+	logrus.Debugf("Rolling back ServiceAccount secret for %s", logKeyFromObject(secret))
+	if err := secretClient.Delete(ctx, secret.Name, metav1.DeleteOptions{}); err != nil {
```
Comment (MbolotSuse) [inline pkg/serviceaccounttoken/secret.go:271]: I think that this error case could used further consideration. Since we don't cleanup after the annotation is set, this could lead to us having a few duplicates (though it's a fixed amount once we get the annotation set, we won't make any more). 

We can address in a future PR if you want since I agree that this is overall better.

**Review by MbolotSuse — COMMENTED** [review]
A few small comments on tests/edge cases.

File: pkg/serviceaccounttoken/secret.go (hunk 6 above)
Comment (bigkevmcd) [reply pkg/serviceaccounttoken/secret.go:287]: I added it to avoid changing the output format, `types.NamespacedName` would be a better option.

This is useful mostly to keep the same ordering in output convention i.e. it's always `namespace/name`.

File: pkg/serviceaccounttoken/secret_test.go [hunk 8]
```diff
@@ -350,193 +342,3 @@ type fakeSecretLister struct {
 func (f *fakeSecretLister) list(namespace string, selector labels.Selector) ([]*corev1.Secret, error) {
 	return f.secrets, f.err
 }
-
-func TestEnsureSecretForServiceAccount_in_parallel(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (bigkevmcd) [reply pkg/serviceaccounttoken/secret_test.go:354]: There's an integration test in `tests/pkg/serviceaccounttoken/mitigation_test.go` which tests this in a more "realistic" way, by spinning up envtest.

Those tests do really fail with appropriate errors because they talk to kube-apiserver rather than the stubbing in the fake client here.

File: pkg/serviceaccounttoken/secret.go (hunk 6 above)
Comment (bigkevmcd) [reply pkg/serviceaccounttoken/secret.go]: I've stripped it in favour of a function that returns `namespace:name`.

**Review by jakefhyde — CHANGES_REQUESTED** [review]

File: pkg/serviceaccounttoken/secret.go (hunk 7 above)
Comment (bigkevmcd) [reply pkg/serviceaccounttoken/secret.go:285]: I think we can do this _after_ we've seen how much of an impact this has.

**Review by jakefhyde — APPROVED** [review]
